src/main/java/core/basesyntax/Rectangle.java

```diff
@@ -27,4 +27,32 @@ public Integer getLength() {
     public String getColor() {
         return color;
     }
+
+    @Override
+    public int hashCode() {
+        int result = 3;
+        result = result * 31 + ((width == null) ? 0 : width.hashCode());
+        result = result * 31 + ((length == null) ? 0 : length.hashCode());
+        result = result * 31 + ((color == null) ? 0 : color.hashCode());
+        return result;
+    }
+
+    @Override
+    public boolean equals(Object object) {
+        if (this == object) {
+            return true;
+        }
+        if (object == null || !object.getClass().equals(Rectangle.class)) {
+            return false;
+        }
+
+        Rectangle rectangle = (Rectangle) object;
+
+        return (width == null && rectangle.width == null
+                || width != null && width.equals(rectangle.width)
+                && length == null && rectangle.length == null
+                || length != null && length.equals(rectangle.length)
+                && color == null && rectangle.color == null
+                || color != null && color.equals(rectangle.color));
```

Comment:
The logical-AND operator ( && ) has higher precedence than the logical-OR operator ( || )
add parenthesis to this statement to ensure the flow is correct

Reply:
It's correct I've made it first with lots oh parenthesis then slowly removed one by one redundant parenthesis. It's working exactly as it should be I've made tests for this with null parameter for one, for two, for both and it works. But I couldadd some parenthesis for better understanding logic?